docs/spec.md

```diff
@@ -170,6 +174,82 @@ Informs the client that channels are no longer available.
 }
 ```
 
+### Client Advertise
+
+- Informs the server about available client channels. Note that the client is only allowed to advertise channels if the server previously declared that it has the `clientPublish` [capability](#server-info).
+
+#### Fields
+
+- `op`: string `"advertise"`
+- `channels`: array of:
+  - `id`: number chosen by the client. The client may reuse ids that have previously been unadvertised.
+  - `topic`: string
+  - `encoding`: string, must be `"json"`
+  - `schemaName`: string
+
+#### Example
+
+```json
+{
+  "op": "advertise",
+  "channels": [
+    {
+      "id": 1,
+      "topic": "foo",
+      "encoding": "json",
+      "schemaName": "ExampleMsg"
+    }
+  ]
+}
+```
+
+### Client Unadvertise
+
+- Informs the server that client channels are no longer available. Note that the client is only allowed to unadvertise channels if the server previously declared that it has the `clientPublish` [capability](#server-info).
+
+#### Fields
+
+- `op`: string `"unadvertise"`
+- `channelIds`: array of number, corresponding to previous [Client Advertise](#client-advertise)
+
+#### Example
+
+```json
+{
+  "op": "unadvertise",
+  "channelIds": [1, 2]
+}
+```
+
+### Client Message Data
+
+- Sends a JSON message from the client to the server. Note that the client is only allowed to publish messages if the server previously declared that it has the `clientPublish` [capability](#server-info).
+
+#### Fields
+
+- `op`: string `"publish"`
+- `channelId`: number. Channel ID corresponding to previous [Client Advertise](#client-advertise)
+- `data`: object. JSON object
```

Comment:
I think this should be a binary message for protocol simplicity. For this one example encoding of JSON a non-binary message works, but most encodings will require binary. It simplifies the server code if we don't need to handle client-published messages with two separate code paths.

Reply:
> For this one example encoding of JSON

As currently written in the PR, JSON is the only allowed encoding. The thinking was that it would add a lot of complexity to ask clients to learn how to encode messages in multiple encodings, and somehow discover which encodings the server would support. How would you recommend we address these issues?

Reply:
> add a lot of complexity to ask clients to learn how to encode messages in multiple encodings

I'm not suggesting clients should support multiple encodings. If a ws-protocol client is doing bag playback of a ROS 1 bag, it makes sense to advertise ros1-encoded topics and publish unprocessed message payloads rather than requiring the client to decode, transform to JSON, then publish. In this example, the client would only support publishing ros1-encoding and if the server did not support that encoding it would not be able to use the publish capability.

This does require the server to advertise which encodings it supports for client publishing. I'm not sure where that slots into the protocol best.